### ISSUE TITLE
[1.16.4] Reward Tables Shifting IDs

### ISSUE THREAD

**Issue by MuteTiefling** [FTBTeam/FTB-Quests#500]
Version: ftb-quests-1604.1.1.15

We're having a lot of trouble with quest reward IDs shifting on our dev server. 

It usually goes something like this:

1.  Add a new reward table
1.  Assign it to various quests
1.  Dev server's changes get commited up to github along with non quest changes being committed elsewhere
1.  Dev server will eventually be brought even with the current commits and brought back up. 

This is the point at which we usually see issues. I'll go into a quest that existed previously and note that the reward table has shifted. For example, every instance of "Epic Loot Reward" in every quest will be replaced with "Rare Loot Reward". Like Reward Table ID 50 will just suddenly get shifted to 51. 

I believe at this point that the actual IDs used by the quests are shifting. The quests themselves keep pointing to ID X, but the reward table will get re-assigned a new ID. 

This only seems to happen when we've added a new reward table, however. The rewards remain stable across reboots, updates from github, or anything else, just so long as there have been no additions to the rewards tables

**Comment by MuteTiefling** [comment]
Bit more information on this. 

We had the server go through several updates without issue, so it`s not inherently a problem with the github sync. Instead, this appears to be more of an issue with freshly created loot tables.

Any time we create a new table it appears to be shifting IDs for older loot tables. Amazingly, it doesn`t seem to be visible in game until after a restart of the server. At which point, the damage becomes obvious as many quests will suddenly have the wrong tables assigned.

**Comment by MuteTiefling** [comment]
I've now received two reports from players running E6 on servers. They haven't modified quests at all, but IDs have shifted for them, just as we were seeing on the dev server.

These are fresh installs in both cases and the quests are all showing correctly on our dev server still. So something has gone wrong during the installation process.

**Comment by MuteTiefling** [comment]
Pack is currently on ftb-quests-1604.1.1.18.jar

We're seeing this for cases where people have updated from older versions and for fresh installs. Many cases are hosted on Bisect, but we can replicate on CreeperHost servers and also have a group of players hosting on a private Linux box that have the same issue.

**Comment by MuteTiefling** [comment]
Well, I think I've realized what the core issue is here.

In the quest files, the loot table IDs are all tied to some index that appears to be generated dynamically.

All of our lot table indexes are from 0 to 52. But if you look in the loot table files themselves, none of them share these values. The range from the hundreds to the thousands.

So I assume on load those files are read in in order and that order is used to generate the index map. But if the system returns those files in a different order (numeric instead of alphabetic, for instance) the index map will differ from what they were when created.

This also explains our primary issue, since creating new loot tables doesn't always make a new table with an ID greater than the existing ones. So as soon as you reload, the index map changes and everything after that ID shifts up one (exactly what I've been seeing every time I've had to fix the loot) 

It seems, then, that the fix for this is to scrap the index map. If a loot table's ID is 2048, then that's the exact string that should appear in the quest file and should be used to load the appropriate loot table.
